### ISSUE TITLE
CONFIG dictionaries stuff 

### ISSUE THREAD

**Issue by huguesdevimeux** [ManimCommunity/manim#7]
We have discussed about the variable `CONFIG `previously on discord. 
What should we do about it ? 

I personally think we should modify at least a little bit of its architecture. it alters the readability a lot. 
One problem with that would be that in changing `CONFIG` we would have to review the entire architecture of the code. 

What do you think?

**Comment by leotrs** [comment]
I agree it's not the most elegant piece of the code. I'd love to hear some suggestions here. I have some experience with the [configparser](https://docs.python.org/3/library/configparser.html) module from the standard library.

**Comment by leotrs** [comment]
A related issue is the `constants.py` file. Both this file and the CONFIG dict are basically ways of tracking hard-coded options. One nice-to-have would be to expose this options and allow the user to change or visualize them more easily. This would be achieved by the config files used by [configparser](https://docs.python.org/3/library/configparser.html).

**Comment by eulertour** [comment]
The CONFIG and constants.py issues can be handled separately, although we should do both. One of the confusing things about constants.py is that some of the variables defined there aren't actually constants (e.g. the media directories).

As for the CONFIG variables in each class I think it'd be cleaner to just add them as keyword arguments. We could leave the final **kwargs argument to accept additional config variables for backwards compatibility. For example Circle.\_\_init\_\_ [here](https://github.com/ManimCommunity/manim/blob/9e3aeb299746fb4c2fe7902e972e9264277c47fe/manimlib/mobject/geometry.py#L313) would be something like
```python
def __init__(self, color=RED, close_new_points=True, anchors_span_full_range=False, **kwargs):
    self.color = color
    self.close_new_points = close_new_points
    self.anchors_span_full_range = anchors_span_full_range
    self.__dict__.update(kwargs)
```

**Comment by leotrs** [comment]
Is it always the case that doing `self.CONFIG['key']` would be equivalent to doing `self.key` instead? for example, are the keys of the CONFIG dictionary iterated over? 

Another thing to think about is: which of the things currently stored in CONFIG should be turned into properties.

**Comment by eulertour** [comment]
> Is it always the case that doing self.CONFIG['key'] would be equivalent to doing self.key instead? for example, are the keys of the CONFIG dictionary iterated over?

Yes, but the code is pretty complicated.
https://github.com/ManimCommunity/manim/blob/master/manimlib/utils/config_ops.py#L23

**Comment by leotrs** [comment]
If that's the case, then I'd vote against converting each key of CONFIG into a member of the class. Because once they're just another member, there is no way of telling apart whether a member is supposed to be one that would go in CONFIG. Said another way, you can't iterate over only the members that are supposed to be CONFIGs without iterating over a bunch of members that are not.

**Comment by PgBiel** [comment]
> If that's the case, then I'd vote against converting each key of CONFIG into a member of the class. Because once they're just another member, there is no way of telling apart whether a member is supposed to be one that would go in CONFIG. Said another way, you can't iterate over only the members that are supposed to be CONFIGs without iterating over a bunch of members that are not.

use __slots__ for this

**Comment by leotrs** [comment]
Oooooooooooooh good one

**Comment by yoshiask** [comment]
As someone who's most familiar with C#, the CONFIG is a big mess to me. There's no indication of which parameters are available and which ones aren't. `kwargs` has the same issue. This may just be a gripe with Python in general, but I think Manim is complicated enough without introducing unknown parameters.

**Comment by yoshiask** [comment]
I prefer @eulertour's proposal
> As for the CONFIG variables in each class I think it'd be cleaner to just add them as keyword arguments. We could leave the final **kwargs argument to accept additional config variables for backwards compatibility.

**Comment by eulertour** [comment]
> If that's the case, then I'd vote against converting each key of CONFIG into a member of the class. Because once they're just another member, there is no way of telling apart whether a member is supposed to be one that would go in CONFIG. Said another way, you can't iterate over only the members that are supposed to be CONFIGs without iterating over a bunch of members that are not.

@leotrs There is no difference between CONFIG variables and instance variables. They're one and the same.

**Comment by leotrs** [comment]
> @leotrs There is no difference between CONFIG variables and instance variables. They're one and the same.

Well shit. Let's use kwargs and be done with it.

**Comment by huguesdevimeux** [comment]
There is a difference that is purely for readability purpose. As @yoshiask said, it put unknows parameters, which is not good in a quite complex library as Manim.
I’m completely in favor of @eulertour ‘point.

**Comment by PgBiel** [comment]
In my opinion, the usage of \_\_slots\_\_ makes it pretty clear to anyone who reads the code which attributes the class may have. So it will turn the code even more readable!

(Also note that \_\_slots\_\_ has memory advantages when many instances of a class are created, but this shouldn't be the main argument since it's pretty much irrelevant... but any improvement in this sense is welcome, of course)

**Comment by leotrs** [comment]
I'm leaning toward `__slots__` now.

**Comment by PgBiel** [comment]
An update on that.

As discussed on Discord, we should use a dataclass model for our classes, using packages such as `dataclasses` (from stdlib) and `attrs`.

Sample code:
```py
from dataclasses import dataclass

@dataclass
class MyCoolClass:
  a = 5
  b = 6

# then
obj = MyCoolClass(a=5, b=6)
# or
obj = MyCoolClass(5, 6)
```

Note that it automatically generates an `__init__` method with each attribute as a parameter (which can be used as a kwarg or not).

Also note that, with the `attrs` package, the code would be the exact same. The only thing that would change would be the name of the decorator:
```py
import attr

@attr.s
class MyCoolClass:
  a = 5
  b = 6

# everything else is the same.
```

**Pros of using `attrs` over `dataclasses`:**
1. **Support for validators** (verification functions) for individual attributes - note, however, that this is possible with `dataclasses`by using the `__postinit__` method;
2. **Support for converters**, which convert the attribute's value straight away. Note that this is the preferred way of converting attributes, since, with `dataclasses`, one would need to convert using `__postinit__`, which means there would be a timespan during which the attribute would have the wrong type, and only on the `__postinit__` call would this be corrected. (Of course, it's not like said timespan is too big though, so we can mostly ignore this)
3. **Automatic creation of `__slots__`**: Really, just write `@attr.s(slots=True)` and it creates the `__slots__`, which gives a memory performance boost when many instances of a class are created. (Otherwise, though, it's a mostly invisible difference)

**Cons of using `attrs`:**
1. **It is another dependency.** However, unlike other manim deps, such as numpy and... God forbid... cairo, it is very unlikely to randomly error during installation, as it is a much simpler module.

What do y'all think?

**Comment by PgBiel** [comment]
^Also, by the way: Worth noting that, if we actually implement this, this could be part of a major manim rework, i.e., working on this would imply reorganizing the code and whatnot. So we must keep this in mind. 

This would likely be positive, of course. But we will need cooperation.

**Comment by leotrs** [comment]
I like the idea of using dataclasses/attrs. I am usually partial to staying within stdlib, but since dataclasses were modeled directly after attrs, I think I'm OK with using attrs.

@yoshiask what do you think about this?

Regarding this being part of a major rework. I propose this: whoever ends up working on this would need to touch a number of files (and should probably wait until we have some minimal tests in place #22), but I suggest that this PR does not include any refactoring beyond changing CONFIG for attrs/dataclasses. In writing this PR, the author will have a better view of what refactoring needs to be done in the future, but we can leave that for later, I think.

**Comment by yoshiask** [comment]
Dataclasses and attrs looks pretty good to me! Unfortunately I can't do it now, but I can start switching everything over to our new structure towards the end of the month, and hopefully by then we'll have CI and stuff set up.

**Comment by leotrs** [comment]
I think we have consensus among everyone who's participated here, except for @eulertour.

**Comment by PgBiel** [comment]
> Dataclasses and attrs looks pretty good to me! Unfortunately I can't do it now, but I can start switching everything over to our new structure towards the end of the month, and hopefully by then we'll have CI and stuff set up.

I'd be more than willing to help here as well!

**Comment by eulertour** [comment]
Do dataclasses/attrs allow for positional arguments to `__init__` and for optional `**kwargs` for example what's used [here](https://github.com/ManimCommunity/manim/blob/master/manimlib/mobject/geometry.py#L281)? I'm reluctant to make such a large change that breaks old code.

**Comment by kilacoda** [comment]
Yes, if you declare the variable with a type but don't assign it a value.
```py
from dataclasses import dataclass

@dataclass
class A:
    a:int
    b:str
    c:list = []
```

**Comment by leotrs** [comment]
Only when declared with a type? Would this make some of our code have type hints while the rest doesn't?

**Comment by kilacoda** [comment]
In python you can only create unassigned variable by providing a type I think. Also, we can always keep the type to be `Any` if we don't know it.

**Comment by leotrs** [comment]
Right. I think I would have to see what manim looks like using dataclasses/attrs before actually committing to its usage. So maybe like, someone can volunteer to take one of the smaller/simpler files and rewrite it using dataclasses. That way we can have concrete pros and cons. Of course this is just a dream, I'm not suggesting that someone spends time on a PR that may or may not be merged.

Alternatively, we just need to think *real* hard about how other parts of the code will be affected by using dataclasses/attrs. (And I wouldn't think of doing this until we have at least a few tests in place...)

**Comment by eulertour** [comment]
@kilacoda I see that that covers positional arguments but does it cover kwargs as well?

**Comment by PgBiel** [comment]
> @kilacoda I see that that covers positional arguments but does it cover kwargs as well?

any parameters in a function can be specified as kwargs unless explicitly told otherwise (feature only available in py 3.8+: pos-only args)

**Comment by PgBiel** [comment]
> I like the idea of using dataclasses/attrs. I am usually partial to staying within stdlib, but since dataclasses were modeled directly after attrs, I think I'm OK with using attrs.

I'd like to follow this course of action: we develop everything using only the stdlib features (`dataclasses`). Then, if we do reach a point where we do need `attrs`, we can switch to it. Also, we may compare memory usage with and without `attrs` + `slots=True`, in order to determine if it is worth it.

**Comment by PgBiel** [comment]
> Do dataclasses/attrs allow for positional arguments to `__init__` and for optional `**kwargs` for example what's used [here](https://github.com/ManimCommunity/manim/blob/master/manimlib/mobject/geometry.py#L281)? I'm reluctant to make such a large change that breaks old code.

For the optional kwargs, one can use inheritance for this: the class should inherit the attributes of its parents. This way we have more type-strict (i.e. easier to typecheck, not actually "strict" because this is python) classes and more predictable models. If it depends on unpredictable parameters, imo it's bad design

**Comment by leotrs** [comment]
> we develop everything using only the stdlib features (dataclasses). Then, if we do reach a point where we do need attrs, we can switch to it.

I'm afraid that if we design/develop everything around dataclasses and get used to those, we might never really feel the need to move on to attrs because, well, we designed it without that need. We may never know what we're missing. It's the problem of unknown unknowns you see.

**Comment by PgBiel** [comment]
Okay well I guess we can do the reverse then? No idea. I mean, it's not like `attrs` has a ton of advantages, but... Well, let's use it then. If we ever feel like changing our mind... we can discuss that later

**Comment by yoshiask** [comment]
> In python you can only create unassigned variable by providing a type I think. Also, we can always keep the type to be `Any` if we don't know it.

I'm actually quite familiar with the data types of many of the variables in Manim (because I had to find a way to project them to C#)

**Comment by leotrs** [comment]
At this point my personal opinion is that @PgBiel and @yoshiask should determine which library to use (between attrs and dataclasses). Why they? Because they assigned this issue to themselves. And I don't see anyone else strongly objecting to either option.

As long as @eulertour's questions have been answered, that is.

**Comment by eulertour** [comment]
My only request is that whatever we settle on, it should still be a drop-in replacement. I wouldn't want to get rid of the CONFIG vars only to introduce another quirk that people still don't understand.

**Comment by PgBiel** [comment]
> At this point my personal opinion is that @PgBiel and @yoshiask should determine which library to use (between attrs and dataclasses). Why they? Because they assigned this issue to themselves. And I don't see anyone else strongly objecting to either option.

I say we go with attrs for now; if needed, we can change our mind later.

> My only request is that whatever we settle on, it should still be a drop-in replacement. I wouldn't want to get rid of the CONFIG vars only to introduce another quirk that people still don't understand.

They (dataclasses) work directly with python attributes. That is, they work very similarly to CONFIG, but using attributes directly (without "digest_config") and allowing for more customization and better type interpretation by IDEs.

**Comment by leotrs** [comment]
@PgBiel looking forward to reviewing that PR! Something tells me to expect a bit more discussion there, but at least we have settled on which library to use.

**Comment by xy-23** [comment]
Hi, I actually post my idea about CONFIG in discord sometime ago, the message is so long that you can go to here https://discord.com/channels/581738731934056449/581738732646957057/701035356950364221
to see it if you like...

The main idea is to use `metaclass`, and after seeing the solution of using `dataclasses/attrs`, I think this is probably a good way(at least better than mine). But the my most concern is whether the attributes defined by `dataclasses/attrs` can be recognized by IDE properly and give the correct auto-completion, because as a non-native English speaker, I can't remember those attributes' name correctly...

Also, I have changed (almost) all the CONIG of mobject into attributes in here sometime ago. https://github.com/xy-23/manim/tree/dev

**Comment by PgBiel** [comment]
> The main idea is to use `metaclass`, but after seeing the solution of using `dataclasses/attrs`, I think this is probably a good way(at least better than mine).

Your metaclass is basically the `attr.s` and `dataclass` decorators, just with less features... so you pretty much converge with our plan

> And my most concern is whether the attributes defined by `dataclasses/attrs` can be recognized by IDE properly and give the correct auto-completion, because as a non-native English speaker, I can't remember those attributes' name correctly...

Yes

> Also, I have changed (almost) all the CONIG of mobject into attributes in here sometime ago. https://github.com/xy-23/manim/tree/dev

I'll take a look!

**Comment by PgBiel** [comment]
Note: work on this should begin after we are done with all other "Initial Cleanup" issues

**Comment by leotrs** [comment]
Can you be more specific? Do you mean you wish the CONFIG stuff was documented before changing it? If so, I suggest not waiting for documentation. Instead, I suggest writing tests for the CONFIG stuff first, and make sure that the dataclass/attr rewrite passes the tests.

**Comment by PgBiel** [comment]
I mean it gets much cleaner if we know the types beforehand... I think all the three projects - dataclasses, typings and documentation - are all kind of connected, and working on one implies on working on all of them

**Comment by leotrs** [comment]
Agreed. And that's exactly why we can't wait until one is done to start the other two. I suggest you delineate what kind of documentation you need for this project and work on it at the same time, on the same PR. You can confer with the people in charge of the docs branch so there are no conflicts.

**Comment by leotrs** [comment]
This is now the only standing issue in the "initial cleanup milestone". @PgBiel, I don't mean to rush you, but do you have any updates? :)

**Comment by PgBiel** [comment]
I'll try to start work on it tomorrow, but I gotta note that I'm gonna be quite busy (especially on Friday) but, again, I'll see what I can do

**Comment by yoshiask** [comment]
Once this is finished and merged, I'll start working on #31

**Comment by leotrs** [comment]
Now that #98 has been merged, it has deprecated/eliminated a bunch of the keys in the CONFIG dictionaries in classes across the codebase. It'd be great to have an update here on what still needs to be done.

**Comment by naveen521kk** [comment]
Is dataclassess introduced in python 3.7 right? Then, doing this would mean that manim won't with python 3.6. If I am right, still Ubuntu Linux has Python 3.6 as default and the latest version of python. Isn't that a problem?

**Comment by leotrs** [comment]
You raise a good point. @PgBiel thoughts?

**Comment by kilacoda** [comment]
We're using `attrs` though and not `dataclasses`.

**Comment by PgBiel** [comment]
correct, kilacoda. We're not going to use `dataclasses`, but `attrs`, which supports most python versions, including python 2

**Comment by naveen521kk** [comment]
> Is dataclassess introduced in python 3.7 right?

Some updates, dataclassess is supported in python 3.6 but not as a default module instead can be installed as a dependency from pip. I had know about this which working poetry and rich causeing me trouble lol.

**Comment by huguesdevimeux** [comment]
But uuh .. isn't the plan to use `attrs` instead of `dataclasses` ? 

> correct, kilacoda. We're not going to use dataclasses, but attrs, which supports most python versions, including python 2

**Comment by eulertour** [comment]
I talked some with Grant and he's now planning to migrate toward using regular classes, so it'd be better for consistency if we did that. Is anyone opposed to that?

**Comment by leotrs** [comment]
As opposed to dataclasses OR attrs?

**Comment by eulertour** [comment]
My takeaway from #271 is that attrs don't fit our use case. Dataclasses might but we don't know what problems we might run into so it'd be safer to stick with vanilla python. If someone is willing to vouch for dataclasses then maybe it'd be worth going back and asking him about it, but my prediction is that it will have some quirk like attrs that makes it difficult to adapt.

**Comment by leotrs** [comment]
In principle I tend to agree. There's some stuff about attrs use of __post__init that doesn't sit well with me, and it seems to me that it will be a problem when moving away from CONFIG.

Cc. @PgBiel

**Comment by cobordism** [comment]
The thing that CONFIG does really nicely is it allows anyone to easily create derived classes with a few default values changed.
```
class myRedText(Text):
    CONFIG = {
        colour: "RED"
    }
``` 
(or whatever the correct syntax is here) 
is a lot simpler than
```
class myRedText(Text):
    def __init__(self, *args, **kwargs):
        c = kwargs.pop("colour", "RED")
        super().__init__(self, args, colour=c, kwargs)
```
or sth like that. (don't criticise my syntax in a github comment :)

My question is, how do dataclasses impact my ability to quickly and easily create such derived classes?

**Comment by leotrs** [comment]
I don't know about dataclasses, but. The example you mention is indeed very convenient for a dev who is defining derived classes. However, this happens extremely rarely compared to the number of times that a user needs to use a class, or read the documentation, or would like to use autocompletion. CONFIG dictionaries make all of these almost impossible.

**Comment by leotrs** [comment]
Also, the latest point brought up is that it may be better to not use either dataclasses or CONFIG. We're waiting for @PgBiel input.

**Comment by cobordism** [comment]
> this happens extremely rarely

As someone who learned python by learning how to make manim videos and reading 3b1b code, I naturally assumed that this happens all the time! Indeed I have been doing a lot of this :D

But I am starting to learn that the 3b1b code is somewhat idiosyncratic and may not be the best guide.

**Comment by leotrs** [comment]
> As someone who learned python by learning how to make manim videos and reading 3b1b code

_yikes_

**Comment by PgBiel** [comment]
> My takeaway from #271 is that attrs don't fit our use case. Dataclasses might but we don't know what problems we might run into so it'd be safer to stick with vanilla python. If someone is willing to vouch for dataclasses then maybe it'd be worth going back and asking him about it, but my prediction is that it will have some quirk like attrs that makes it difficult to adapt.

What is this quirk you speak about? 

But, vanilla classes could work fine if we do it right. If you have any ideas about this, you could make some model or prototype or something so that I can better understand what you have in mind

**Comment by eulertour** [comment]
@PgBiel The issues at the top of #271 
https://github.com/ManimCommunity/manim/pull/271#issue-466912315

**Comment by PgBiel** [comment]
Regarding the first issue, I'm yet to come upon it, so I'll take a look later

Regarding the second, I'm pretty sure there are ways to bypass that; specifying types again is only necessary because of how python handles types in class attrs: it adds them to `clazz.__annotations__`, and, if you don't specify a type, the attribute is not stored there. However, if we don't use the automatic type to annotation feature for those cases, simply using `a = attr.ib(b)` instead of the fancy `a: fulltype = b` might have it automatically infer that the type is the same as that of the attribute in the superclass (I'm yet to test this, but I believe this could work).

Anyway, we could work with vanilla classes. However, I believe we should keep attrs for specific classes. We should use it where convenient, and maybe in important classes or something (to be discussed later).

**Comment by leotrs** [comment]
It seems to me that a lot of the confusion here was moving from the current implementation using `CONFIG` dicts directly to attrs/dataclasses. 

Now I'm more inclined to first move to vanilla classes ASAP, and then decide which classes to implement using attrs. Now that #620 is merged, I can work on removing `CONFIG` entirely and using vanilla classes. Once that's done we can revisit which classes need to move to attrs.

Yes, it's more work, but I'm willing to do it.

**Comment by cobordism** [comment]
> Now I'm more inclined to first move to vanilla classes ASAP
This is now done in PR #783 
The plan is to merge that just after release 0.1.1 - so that it will be included in the subsequent 0.2.0 but we have a whole month to clean up.

However, once 783 is merged we have to address the question of how we want to refactor this going forward. dataclasses? fastcore? ...
We may also need a different approach for those CONFIGs that used to be part of mobjects that a user might instantiate vs the CONFIGS that were in Scenes that a user was meant to subclass (like GraphScene for example). The former only requires keyword arguments to the constructor, the latter might require a custom `__init__()`. 

...Of course, why a GraphScene is a scene at all, as opposed to a Graph being a mobject is another question entirely ...

**Comment by leotrs** [comment]
> we have to address the question of how we want to refactor this going forward. dataclasses? fastcore? ...

The question is: what do our classes need right now? I'm not certain we necessarily have a problem that needs to be fixed right now. (Other than the following point you make, of course...)

> We may also need a different approach for those CONFIGs that used to be part of mobjects that a user might instantiate vs the CONFIGS that were in Scenes that a user was meant to subclass (like GraphScene for example). The former only requires keyword arguments to the constructor, the latter might require a custom __init__().

I'm fairly certain that most (if not all) of the CONFIG keys of GraphScene can just be replaced by setting attributes at the top of the construct function. The code
```python
class MyScene(GraphScene):
    CONFIG = {"x_min": -1}
    def construct(self):
        self.setup_axes()
```
Can be replaced with 
```python
class MyScene(GraphScene):
    def construct(self):
        self.x_min = -1
        self.setup_axes()
```
In fact, we could encourage use of `Scene.setup()` for this purpose.

> ...Of course, why a GraphScene is a scene at all, as opposed to a Graph being a mobject is another question entirely ...

I think we must answer this question before choosing if we need to migrate to any other class and/or how and when. What is the concrete problem that needs to be addressed after we remove CONFIG once and for all?

**Comment by cobordism** [comment]
closed by PR #783 

For discussions on future refactoring, please open new dedicated issues.
